cwlVersion: v1.0
class: Workflow


requirements:
  - class: SubworkflowFeatureRequirement
  - class: StepInputExpressionRequirement
  - class: InlineJavascriptRequirement
  - class: MultipleInputFeatureRequirement


'sd:upstream':
  sc_rnaseq_sample:
  - "single-cell-preprocess-cellranger.cwl"
  - "cellranger-aggr.cwl"
  - "https://github.com/datirium/workflows/workflows/single-cell-preprocess-cellranger.cwl"
  - "https://github.com/datirium/workflows/workflows/cellranger-aggr.cwl"
  altanalyze_prepare_genome_sample:
  - "altanalyze-prepare-genome.cwl"
  - "https://github.com/datirium/workflows/workflows/altanalyze-prepare-genome.cwl"


inputs:

  alias_:
    type: string
    label: "Experiment short name/Alias"
    sd:preview:
      position: 1

  genome_data:
    type: Directory
    label: "AltAnalyze Prepare Genome Experiment"
    doc: "Ensembl database from the altanalyze-prepare-genome.cwl pipeline"
    'sd:upstreamSource': "altanalyze_prepare_genome_sample/genome_data"
    'sd:localLabel': true

  filtered_feature_bc_matrix_h5:
    type: File[]
    label: "scRNA-Seq Cell Ranger Count or Aggregate Experiments"
    doc: "Filtered feature-barcode matrices in HDF5 format from cellranger count or aggr results"
    'sd:upstreamSource': "sc_rnaseq_sample/filtered_feature_bc_matrix_h5"
    'sd:localLabel': true

  exclude_cell_cycle:
    type:
    - "null"
    - type: enum
      symbols:
      - "no"
      - "yes"
      - "strict"
    default: "no"
    label: "Exclude cell cycle"
    doc: "Exclude cell cycle"
    'sd:layout':
      advanced: true

  remove_outliers:
    type: boolean?
    default: true
    label: "Remove outliers"
    doc: "Remove outliers"
    'sd:layout':
      advanced: true

  restrict_by:
    type:
    - "null"
    - type: enum
      symbols:
      - "None"
      - "protein_coding"
    default: "None"
    label: "Restrict by"
    doc: "Restrict by"
    'sd:layout':
      advanced: true

  downsample:
    type: int?
    default: 5000
    label: "Downsample to"
    doc: "Downsample to"
    'sd:layout':
      advanced: true

  marker_pearson_cutoff:
    type: float?
    default: 0.3
    label: "Marker Pearson cutoff"
    doc: "Marker Pearson cutoff"
    'sd:layout':
      advanced: true

  max_cluster_count:
    type: int?
    default: 20
    label: "Maximum number of clusters to be detected"
    doc: "Maximum number of clusters to be detected"
    'sd:layout':
      advanced: true

  resolution:
    type: float?
    default: 0.4
    label: "Clustering resolution"
    doc: "Clustering resolution"
    'sd:layout':
      advanced: true


outputs:

  compressed_icgs_data_folder:
    type: File
    outputSource: compress_icgs_data_folder/compressed_folder
    label: "Compressed folder with AltAnalyze ICGS results"
    doc: |
      Compressed folder with AltAnalyze ICGS results

  compressed_html_data_folder:
    type: File
    outputSource: compress_html_data_folder/compressed_folder
    label: "Compressed folder with CellBrowser formatted results"
    doc: |
      Compressed folder with CellBrowser formatted results

  html_data_folder:
    type: Directory
    outputSource: cellbrowser_build/html_data
    label: "Folder with not compressed CellBrowser formatted results"
    doc: |
      Folder with not compressed CellBrowser formatted results

  cellbrowser_report:
    type: File
    outputSource: cellbrowser_build/index_html_file
    label: "CellBrowser formatted AltAnalyze ICGS report"
    doc: |
      CellBrowser formatted AltAnalyze ICGS report
    'sd:visualPlugins':
    - linkList:
        tab: 'Overview'
        target: "_blank"

  icgs_marker_heatmap_file:
    type: File
    outputSource: altanalyze_icgs/marker_heatmap_file
    label: "AltAnalyze ICGS Marker Gene Heatmap"
    doc: |
      AltAnalyze ICGS Marker Gene Heatmap in txt format

  icgs_marker_heatmap_png:
    type: File
    outputSource: altanalyze_icgs/marker_heatmap_png
    label: "AltAnalyze ICGS Marker Gene Heatmap"
    doc: |
      AltAnalyze ICGS Marker Gene Heatmap in png format
    'sd:visualPlugins':
    - image:
        tab: 'Plots'
        Caption: 'Marker Gene Heatmap'

  icgs_annotation_metadata_file:
    type: File
    outputSource: altanalyze_icgs/annotation_metadata_file
    label: "Annotation metadata"
    doc: |
      Annotation metadata file.
      Get it from ICGS-NMF/cellbrowser/FinalGroups-CellTypesFull.txt

  icgs_expression_matrix_file:
    type: File
    outputSource: altanalyze_icgs/expression_matrix_file
    label: "Expression matrix"
    doc: |
      Expression matrix file.
      Get it from ICGS-NMF/cellbrowser/exp.cellbrowser.txt

  altanalyze_icgs_stdout_log:
    type: File
    outputSource: altanalyze_icgs/stdout_log
    label: stdout log generated by altanalyze icgs
    doc: |
      stdout log generated by altanalyze icgs

  altanalyze_icgs_stderr_log:
    type: File
    outputSource: altanalyze_icgs/stderr_log
    label: stderr log generated by altanalyze icgs
    doc: |
      stderr log generated by altanalyze icgs


steps:

  altanalyze_icgs:
    run: ../tools/altanalyze-icgs.cwl
    in:
      genome_data: genome_data
      feature_bc_matrices_h5: filtered_feature_bc_matrix_h5
      exclude_cell_cycle: exclude_cell_cycle
      remove_outliers: remove_outliers
      restrict_by: restrict_by
      downsample: downsample
      marker_pearson_cutoff: marker_pearson_cutoff
      max_cluster_count: max_cluster_count
      resolution: resolution
    out:
    - icgs_data
    - expression_matrix_file
    - annotation_metadata_file
    - cell_coordinates_file
    - cluster_marker_genes_file
    - marker_heatmap_file
    - marker_heatmap_png
    - stdout_log
    - stderr_log

  compress_icgs_data_folder:
    run: ../tools/tar-compress.cwl
    in:
      folder_to_compress: altanalyze_icgs/icgs_data
    out:
    - compressed_folder

  cellbrowser_build:
    run: ../tools/cellbrowser-build-altanalyze.cwl
    in:
      expression_matrix_file: altanalyze_icgs/expression_matrix_file
      annotation_metadata_file: altanalyze_icgs/annotation_metadata_file
      cell_coordinates_file: altanalyze_icgs/cell_coordinates_file
      cluster_marker_genes_file: altanalyze_icgs/cluster_marker_genes_file
    out:
    - html_data
    - index_html_file

  compress_html_data_folder:
    run: ../tools/tar-compress.cwl
    in:
      folder_to_compress: cellbrowser_build/html_data
    out:
    - compressed_folder


$namespaces:
  s: http://schema.org/

$schemas:
- https://github.com/schemaorg/schemaorg/raw/main/data/releases/11.01/schemaorg-current-http.rdf

s:name: "AltAnalyze ICGS"
label: "AltAnalyze ICGS"
s:alternateName: "Runs iterative clustering and guide-gene selection for Cell Ranger Count Gene Expression or Cell Ranger Aggregate experiments"

s:downloadUrl: https://raw.githubusercontent.com/datirium/workflows/master/workflows/altanalyze-icgs.cwl
s:codeRepository: https://github.com/datirium/workflows
s:license: http://www.apache.org/licenses/LICENSE-2.0

s:isPartOf:
  class: s:CreativeWork
  s:name: Common Workflow Language
  s:url: http://commonwl.org/

s:creator:
- class: s:Organization
  s:legalName: "Cincinnati Children's Hospital Medical Center"
  s:location:
  - class: s:PostalAddress
    s:addressCountry: "USA"
    s:addressLocality: "Cincinnati"
    s:addressRegion: "OH"
    s:postalCode: "45229"
    s:streetAddress: "3333 Burnet Ave"
    s:telephone: "+1(513)636-4200"
  s:logo: "https://www.cincinnatichildrens.org/-/media/cincinnati%20childrens/global%20shared/childrens-logo-new.png"
  s:department:
  - class: s:Organization
    s:legalName: "Allergy and Immunology"
    s:department:
    - class: s:Organization
      s:legalName: "Barski Research Lab"
      s:member:
      - class: s:Person
        s:name: Michael Kotliar
        s:email: mailto:misha.kotliar@gmail.com
        s:sameAs:
        - id: http://orcid.org/0000-0002-6486-3898


doc: |
  AltAnalyze ICGS
  ===============