cwlVersion: v1.0
class: Workflow


requirements:
  - class: SubworkflowFeatureRequirement
  - class: StepInputExpressionRequirement
  - class: InlineJavascriptRequirement
  - class: MultipleInputFeatureRequirement


'sd:upstream':
  genome_indices:
    - "genome-indices.cwl"


inputs:

  alias:
    type: string
    label: "Experiment short name/Alias"
    sd:preview:
      position: 1

  genome_fasta_file:
    type: File
    format: "http://edamontology.org/format_1929"
    label: "Genome type"
    doc: "Reference genome FASTA file that includes all chromosomes"
    'sd:upstreamSource': "genome_indices/fasta_output"
    'sd:localLabel': true

  annotation_gtf_file:
    type: File
    format: "http://edamontology.org/format_2306"
    label: "Genome type"
    doc: "GTF annotation file that includes refGene and mitochondrial DNA annotations"
    'sd:upstreamSource': "genome_indices/annotation_gtf"
    'sd:localLabel': true

  threads:
    type: int?
    default: 4
    label: "Number of threads"
    doc: "Number of threads for those steps that support multithreading"
    'sd:layout':
      advanced: true

  memory_limit:
    type: int?
    default: 30
    label: "Maximum memory used (GB)"
    doc: "Maximum memory used (GB). The same will be applied to virtual memory"
    'sd:layout':
      advanced: true


outputs:

  indices_folder:
    type: Directory
    outputSource: cellranger_mkref/indices_folder
    label: Cell Ranger genome indices
    doc: |
      Cell Ranger generated genome indices folder

  stdout_log:
    type: File
    outputSource: cellranger_mkref/stdout_log
    label: stdout log generated by cellranger mkref
    doc: |
      stdout log generated by cellranger mkref

  stderr_log:
    type: File
    outputSource: cellranger_mkref/stderr_log
    label: stderr log generated by cellranger mkref
    doc: |
      stderr log generated by cellranger mkref


steps:

  cellranger_mkref:
    run: ../tools/cellranger-mkref.cwl
    in:
      genome_fasta_file: genome_fasta_file
      annotation_gtf_file: annotation_gtf_file
      threads: threads
      memory_limit: memory_limit
    out:
    - indices_folder
    - stdout_log
    - stderr_log


$namespaces:
  s: http://schema.org/

$schemas:
- https://github.com/schemaorg/schemaorg/raw/main/data/releases/11.01/schemaorg-current-http.rdf

s:name: "Single Cell Reference Genome"
label: "Single Cell Reference Genome"
s:alternateName: "Single Cell Reference Genome"

s:downloadUrl: https://raw.githubusercontent.com/datirium/workflows/master/workflows/cellranger-mkref.cwl
s:codeRepository: https://github.com/datirium/workflows
s:license: http://www.apache.org/licenses/LICENSE-2.0

s:isPartOf:
  class: s:CreativeWork
  s:name: Common Workflow Language
  s:url: http://commonwl.org/

s:creator:
- class: s:Organization
  s:legalName: "Cincinnati Children's Hospital Medical Center"
  s:location:
  - class: s:PostalAddress
    s:addressCountry: "USA"
    s:addressLocality: "Cincinnati"
    s:addressRegion: "OH"
    s:postalCode: "45229"
    s:streetAddress: "3333 Burnet Ave"
    s:telephone: "+1(513)636-4200"
  s:logo: "https://www.cincinnatichildrens.org/-/media/cincinnati%20childrens/global%20shared/childrens-logo-new.png"
  s:department:
  - class: s:Organization
    s:legalName: "Allergy and Immunology"
    s:department:
    - class: s:Organization
      s:legalName: "Barski Research Lab"
      s:member:
      - class: s:Person
        s:name: Michael Kotliar
        s:email: mailto:misha.kotliar@gmail.com
        s:sameAs:
        - id: http://orcid.org/0000-0002-6486-3898


doc: |
  Single Cell Reference Genome
  ============================