cwlVersion: v1.0
class: Workflow
requirements:
- class: SubworkflowFeatureRequirement
- class: StepInputExpressionRequirement
- class: InlineJavascriptRequirement
- class: MultipleInputFeatureRequirement
sd:upstream:
  sc_arc_sample:
  - cellranger-arc-count.cwl
  genome_indices:
  - cellranger-mkref.cwl
inputs:
  alias:
    type: string
    label: Analysis name
    sd:preview:
      position: 1
  rna_molecule_info_h5:
    type: File[]
    label: Cell Ranger RNA+ATAC Sample
    doc: |
      Any "Cell Ranger RNA+ATAC Sample"
      that produces both gene expression
      and chromatin accessibility data
      from a single 10x Genomics library
    sd:upstreamSource: sc_arc_sample/rna_molecule_info_h5
    sd:localLabel: true
  gem_well_labels:
    type: string[]
    sd:upstreamSource: sc_arc_sample/alias
  atac_fragments_file_from_count:
    type: File[]
    secondaryFiles:
    - .tbi
    sd:upstreamSource: sc_arc_sample/atac_fragments_file
  barcode_metrics_report:
    type: File[]
    sd:upstreamSource: sc_arc_sample/barcode_metrics_report
  indices_folder:
    type: Directory
    label: Cell Ranger Reference Sample
    doc: |
      Any "Cell Ranger Reference Sample" that
      builds a reference genome package of a
      selected species for quantifying gene
      expression and chromatin accessibility.
      This sample can be obtained from "Cell
      Ranger Reference (RNA, ATAC, RNA+ATAC)"
      pipeline.
    sd:upstreamSource: genome_indices/arc_indices_folder
    sd:localLabel: true
  annotation_gtf_file:
    type: File
    sd:upstreamSource: genome_indices/genome_indices/annotation_gtf
  memory_limit:
    type: int?
    default: 20
    sd:upstreamSource: genome_indices/memory_limit
  normalization_mode:
    type:
    - 'null'
    - type: enum
      symbols:
      - none
      - depth
    default: none
    label: Library depth normalization
    doc: "When \"depth\" normalization is \nselected, subsample reads from\nhigher-depth GEM wells until we\nequalize the 1) median number\nof unique fragments per cell for\neach ATAC library, 2) mean number\nof reads that are confidently\nmapped to the transcriptome per\ncell for each gene expression\nlibrary.\n"
    sd:layout:
      advanced: true
  threads:
    type:
    - 'null'
    - type: enum
      symbols:
      - '1'
      - '2'
      - '3'
      - '4'
      - '5'
      - '6'
    default: '4'
    label: Cores/CPUs
    doc: |
      Parallelization parameter to define the
      number of cores/CPUs that can be utilized
      simultaneously.
      Default: 4
    sd:layout:
      advanced: true
outputs:
  web_summary_report:
    type: File
    outputSource: aggregate_counts/web_summary_report
    label: Cell Ranger Summary
    doc: |
      Report generated by Cell Ranger
    sd:visualPlugins:
    - linkList:
        tab: Overview
        target: _blank
  cellbrowser_report:
    type: File
    outputSource: cellbrowser_build/index_html_file
    label: UCSC Cell Browser
    doc: |
      UCSC Cell Browser HTML index file
    sd:visualPlugins:
    - linkList:
        tab: Overview
        target: _blank
  metrics_summary_report:
    type: File
    outputSource: aggregate_counts/metrics_summary_report
    label: Run summary metrics
    doc: |
      Cell Ranger generated run summary
      metrics in CSV format
  aggregation_metadata:
    type: File
    outputSource: aggregate_counts/aggregation_metadata
    label: Aggregation metadata
    doc: |
      Aggregation metadata file
      in CSV format
  grouping_data:
    type: File
    outputSource: aggregate_counts/grouping_data
    label: Example of datasets grouping
    doc: |
      Example of TSV file to define datasets grouping
  filtered_feature_bc_matrix_folder:
    type: File
    outputSource: compress_filtered_feature_bc_matrix_folder/compressed_folder
    label: Filtered feature barcode matrix, MEX
    doc: |
      Filtered feature barcode matrix stored
      as a CSC sparse matrix in MEX format.
      The rows consist of all the gene and
      peak features concatenated together
      (identical to raw feature barcode
      matrix) and the columns are restricted
      to those barcodes that are identified
      as cells.
  filtered_feature_bc_matrix_h5:
    type: File
    outputSource: aggregate_counts/filtered_feature_bc_matrix_h5
    label: Filtered feature barcode matrix, HDF5
    doc: |
      Filtered feature barcode matrix stored
      as a CSC sparse matrix in hdf5 format.
      The rows consist of all the gene and
      peak features concatenated together
      (identical to raw feature barcode
      matrix) and the columns are restricted
      to those barcodes that are identified
      as cells.
  raw_feature_bc_matrices_folder:
    type: File
    outputSource: compress_raw_feature_bc_matrices_folder/compressed_folder
    label: Raw feature barcode matrix, MEX
    doc: |
      Raw feature barcode matrix stored as
      a CSC sparse matrix in MEX format.
      The rows consist of all the gene and
      peak features concatenated together
      and the columns consist of all observed
      barcodes with non-zero signal for
      either ATAC or gene expression.
  raw_feature_bc_matrices_h5:
    type: File
    outputSource: aggregate_counts/raw_feature_bc_matrices_h5
    label: Raw feature barcode matrix, HDF5
    doc: |
      Raw feature barcode matrix stored as
      a CSC sparse matrix in hdf5 format.
      The rows consist of all the gene and
      peak features concatenated together
      and the columns consist of all observed
      barcodes with non-zero signal for
      either ATAC or gene expression.
  secondary_analysis_report_folder:
    type: File
    outputSource: compress_secondary_analysis_report_folder/compressed_folder
    label: Secondary analysis
    doc: |
      Various secondary analyses that
      utilize the ATAC, RNA data, and
      their linkage: dimensionality
      reduction and clustering results
      for the ATAC and RNA data,
      differential expression, and
      differential accessibility for all
      clustering results above and linkage
      between ATAC and RNA data.
  loupe_browser_track:
    type: File
    outputSource: aggregate_counts/loupe_browser_track
    label: Loupe Browser visualization
    doc: |
      Loupe Browser visualization file
      with all the analysis outputs
  atac_fragments_file:
    type: File
    outputSource: aggregate_counts/atac_fragments_file
    label: ATAC fragments
    doc: |
      Count and barcode information for
      every ATAC fragment observed in
      the experiment in TSV format.
  atac_peaks_bed_file:
    type: File
    outputSource: aggregate_counts/atac_peaks_bed_file
    label: ATAC peaks
    doc: |
      Locations of open-chromatin regions
      identified in this sample. These
      regions are referred to as "peaks".
  atac_peak_annotation_file:
    type: File
    outputSource: aggregate_counts/atac_peak_annotation_file
    label: ATAC peaks annotations
    doc: |
      Annotations of peaks based on
      genomic proximity alone. Note,
      that these are not functional
      annotations and they do not make
      use of linkage with RNA data.
  aggregate_counts_stdout_log:
    type: File
    outputSource: aggregate_counts/stdout_log
    label: Output log, cellranger-arc aggr step
    doc: |
      stdout log generated by cellranger-arc aggr
  aggregate_counts_stderr_log:
    type: File
    outputSource: aggregate_counts/stderr_log
    label: Error log, cellranger-arc aggr step
    doc: |
      stderr log generated by cellranger-arc aggr
  html_data_folder:
    type: Directory
    outputSource: cellbrowser_build/html_data
    label: UCSC Cell Browser data
    doc: |
      Directory with UCSC Cell Browser
      data
steps:
  aggregate_counts:
    run: ../tools/cellranger-arc-aggr.cwl
    in:
      atac_fragments_file_from_count: atac_fragments_file_from_count
      barcode_metrics_report: barcode_metrics_report
      rna_molecule_info_h5: rna_molecule_info_h5
      gem_well_labels: gem_well_labels
      indices_folder: indices_folder
      normalization_mode: normalization_mode
      threads:
        source: threads
        valueFrom: $(parseInt(self))
      memory_limit: memory_limit
      virt_memory_limit: memory_limit
    out:
    - web_summary_report
    - metrics_summary_report
    - atac_fragments_file
    - atac_peaks_bed_file
    - atac_peak_annotation_file
    - secondary_analysis_report_folder
    - filtered_feature_bc_matrix_folder
    - filtered_feature_bc_matrix_h5
    - raw_feature_bc_matrices_folder
    - raw_feature_bc_matrices_h5
    - aggregation_metadata
    - grouping_data
    - loupe_browser_track
    - stdout_log
    - stderr_log
  compress_filtered_feature_bc_matrix_folder:
    run: ../tools/tar-compress.cwl
    in:
      folder_to_compress: aggregate_counts/filtered_feature_bc_matrix_folder
    out:
    - compressed_folder
  compress_raw_feature_bc_matrices_folder:
    run: ../tools/tar-compress.cwl
    in:
      folder_to_compress: aggregate_counts/raw_feature_bc_matrices_folder
    out:
    - compressed_folder
  compress_secondary_analysis_report_folder:
    run: ../tools/tar-compress.cwl
    in:
      folder_to_compress: aggregate_counts/secondary_analysis_report_folder
    out:
    - compressed_folder
  cellbrowser_build:
    run: ../tools/cellbrowser-build-cellranger-arc.cwl
    in:
      secondary_analysis_report_folder: aggregate_counts/secondary_analysis_report_folder
      filtered_feature_bc_matrix_folder: aggregate_counts/filtered_feature_bc_matrix_folder
      annotation_gtf_file: annotation_gtf_file
      aggregation_metadata: aggregate_counts/aggregation_metadata
    out:
    - html_data
    - index_html_file
label: Cell Ranger Aggregate (RNA+ATAC)
doc: |-
  Cell Ranger Aggregate (RNA+ATAC)

  Combines outputs from multiple runs of “Cell Ranger Count (RNA+ATAC)”
  pipeline. The results of this workflow are primarily used in
  “Single-Cell Multiome ATAC and RNA-Seq Filtering Analysis” pipeline.
sd:version: 100
