cwlVersion: v1.0
class: Workflow
requirements:
- class: SubworkflowFeatureRequirement
- class: StepInputExpressionRequirement
- class: InlineJavascriptRequirement
- class: MultipleInputFeatureRequirement
inputs:
  alias:
    type: string
    label: Experiment short name/alias
    sd:preview:
      position: 1
  genome_fasta_file:
    type: File
    label: Genome FASTA file. Hard/soft-masked files are not allowed.
    doc: |
      Genome FASTA file. Hard/soft-masked files are not allowed.
      For example:
      https://ftp.ensembl.org/pub/current_fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
  annotation_gtf_file:
    type: File
    label: GTF annotation file. Should include gene_biotype/transcript_biotype fields.
    doc: |
      GTF annotation file. Should include gene_biotype/transcript_biotype fields.
      For example:
      https://ftp.ensembl.org/pub/current_gtf/homo_sapiens/Homo_sapiens.GRCh38.108.gtf.gz
  memory_limit:
    type: int?
    default: 20
    label: Maximum memory used (GB)
    doc: |
      Maximum memory used (GB).
    sd:layout:
      advanced: true
  threads:
    type:
    - 'null'
    - type: enum
      symbols:
      - '1'
      - '2'
      - '3'
      - '4'
      - '5'
      - '6'
    default: '4'
    label: Cores/CPUs
    doc: |
      Parallelization parameter to define the
      number of cores/CPUs that can be utilized
      simultaneously.
      Default: 4
    sd:layout:
      advanced: true
outputs:
  indices_folder:
    type: Directory
    outputSource: cellranger_mkvdjref/indices_folder
    label: Cell Ranger V(D)J genome indices
    doc: |
      Cell Ranger V(D)J-compatible reference folder.
      This folder will include V(D)J segment FASTA file.
  stdout_log:
    type: File
    outputSource: cellranger_mkvdjref/stdout_log
    label: stdout log generated by cellranger mkvdjref
    doc: |
      stdout log generated by cellranger mkvdjref
  stderr_log:
    type: File
    outputSource: cellranger_mkvdjref/stderr_log
    label: stderr log generated by cellranger mkvdjref
    doc: |
      stderr log generated by cellranger mkvdjref
steps:
  extract_fasta:
    run: ../tools/extract-7z.cwl
    in:
      file_to_extract: genome_fasta_file
      output_filename:
        default: annotation.fasta
    out:
    - extracted_file
  extract_gtf:
    run: ../tools/extract-7z.cwl
    in:
      file_to_extract: annotation_gtf_file
      output_filename:
        default: annotation.gtf
    out:
    - extracted_file
  cellranger_mkvdjref:
    run: ../tools/cellranger-mkvdjref.cwl
    in:
      genome_fasta_file: extract_fasta/extracted_file
      annotation_gtf_file: extract_gtf/extracted_file
      threads:
        source: threads
        valueFrom: $(parseInt(self))
      memory_limit: memory_limit
      output_folder_name:
        default: cellranger_vdj_ref
    out:
    - indices_folder
    - stdout_log
    - stderr_log
label: Cell Ranger Reference (VDJ)
doc: |-
  Cell Ranger Reference (VDJ)

  Builds a reference genome of a selected species for V(D)J
  contigs assembly and clonotype calling. The results of this
  workflow are used in “Cell Ranger Count (RNA+VDJ)” pipeline.
sd:version: 100
