cwlVersion: v1.1
class: Workflow


requirements:
  - class: SubworkflowFeatureRequirement
  - class: StepInputExpressionRequirement
  - class: MultipleInputFeatureRequirement
  - class: InlineJavascriptRequirement
    expressionLib:
    - var split_numbers = function(line) {
          let splitted_line = line?line.split(/[\s,]+/).map(parseFloat):null;
          return (splitted_line && !!splitted_line.length)?splitted_line:null;
      };


"sd:upstream":
  sc_arc_sample:
  - "cellranger-arc-count.cwl"
  - "cellranger-arc-aggr.cwl"


inputs:

  alias:
    type: string
    label: "Analysis name"
    sd:preview:
      position: 1

  filtered_feature_bc_matrix_folder:
    type: File
    label: "Cell Ranger ARC Sample"
    doc: |
      Any "Cell Ranger ARC Sample" that produces
      compressed folder with feature-barcode
      matrix in MEX format, ATAC fragments file
      in TSV format, and optional aggregation
      metadata file in TSV/CSV format.
    "sd:upstreamSource": "sc_arc_sample/filtered_feature_bc_matrix_folder"
    "sd:localLabel": true

  atac_fragments_file:
    type: File
    secondaryFiles:
    - .tbi
    "sd:upstreamSource": "sc_arc_sample/atac_fragments_file"

  aggregation_metadata:
    type: File?
    "sd:upstreamSource": "sc_arc_sample/aggregation_metadata"

  annotation_gtf_file:
    type: File
    "sd:upstreamSource": "sc_arc_sample/genome_indices/genome_indices/annotation_gtf"

  chrom_length_file:
    type: File
    "sd:upstreamSource": "sc_arc_sample/genome_indices/chrom_length_file"

  blacklist_regions_file:
    type:
    - "null"
    - type: enum
      symbols:
      - "hg19"
      - "hg38"
      - "mm10"
    "sd:upstreamSource": "sc_arc_sample/genome_indices/genome_indices/genome"

  grouping_data:
    type: File?
    label: "Datasets grouping (optional)"
    doc: |
      If the selected "Cell Ranger ARC
      Sample" includes multiple aggregated
      datasets, each dataset can be assigned
      to a separate group by providing a
      TSV/CSV file with "library_id" and
      "condition" columns. Obtain this file
      from the "aggregation_metadata.csv"
      output generated by "Cell Ranger ARC
      Sample" and accessible on the "Files"
      tab. Remove all columns except the
      "library_id". Add the group names for
      each dataset in a separate column
      named "condition".

  barcodes_data:
    type: File?
    label: "Selected cell barcodes (optional)"
    doc: |
      A TSV/CSV file to optionally prefilter
      the single cell data by including only
      the cells with the selected barcodes.
      The provided file should include at
      least one column named "barcode", with
      one cell barcode per line. All other
      columns, except for "barcode", will be
      added to the single cell metadata loaded
      from "Cell Ranger ARC Sample" and can be
      utilized in the current or future steps
      of analysis.

  remove_doublets:
    type:
    - type: enum
      symbols:
      - "Based on either RNA or ATAC"
      - "Based on RNA"
      - "Based on ATAC"
      - "Based on both RNA and ATAC"
      - "Do not remove"
    default: "Do not remove"
    label: "Doublets removal"
    doc: |
      Quality control filtering parameter
      to remove cells identified as doublets.
      Depending on the selected option,
      doublets can be detected and removed
      based on only gene expression or
      chromatin accessibility data, or their
      combination (union or intersection).
      Default: do not remove
    "sd:layout":
      advanced: true

  minimum_umis:
    type: string?
    default: "500"
    label: "Minimum number of transcripts per cell"
    doc: |
      Quality control filtering threshold
      to exclude from the analysis all
      cells with the number of transcripts
      smaller than the provided value.
      If the selected "Cell Ranger ARC
      Sample" includes multiple aggregated
      datasets, each of them can be filtered
      independently by providing comma or
      space-separated list of filtering
      thresholds. The order and number of
      the specified values need to match
      with the datasets order from the
      "aggregation_metadata.csv" output
      generated by "Cell Ranger ARC Sample"
      and accessible on the "Files" tab.
      Default: 500
    "sd:layout":
      advanced: true

  minimum_genes:
    type: string?
    default: "250"
    label: "Minimum number of genes per cell"
    doc: |
      Quality control filtering threshold
      to exclude from the analysis all
      cells with the number of expressed
      genes smaller than the provided value.
      If the selected "Cell Ranger ARC
      Sample" includes multiple aggregated
      datasets, each of them can be filtered
      independently by providing comma or
      space-separated list of filtering
      thresholds. The order and number of
      the specified values need to match
      with the datasets order from the
      "aggregation_metadata.csv" output
      generated by "Cell Ranger ARC Sample"
      and accessible on the "Files" tab.
      Default: 250
    "sd:layout":
      advanced: true

  maximum_genes:
    type: string?
    default: "5000"
    label: "Maximum number of genes per cell"
    doc: |
      Quality control filtering threshold
      to exclude from the analysis all
      cells with the number of expressed
      genes bigger than the provided value.
      If the selected "Cell Ranger ARC
      Sample" includes multiple aggregated
      datasets, each of them can be filtered
      independently by providing comma or
      space-separated list of filtering
      thresholds. The order and number of
      the specified values need to match
      with the datasets order from the
      "aggregation_metadata.csv" output
      generated by "Cell Ranger ARC Sample"
      and accessible on the "Files" tab.
      Default: 5000
    "sd:layout":
      advanced: true

  mito_pattern:
    type: string?
    default: "^mt-|^MT-"
    label: "Mitochondrial genes pattern"
    doc: |
      Regex pattern to identify mitochondrial
      genes based on their names.
      Default: "^mt-|^MT-"
    "sd:layout":
      advanced: true

  maximum_mito_perc:
    type: float?
    default: 5
    label: "Maximum mitochondrial percentage per cell"
    doc: |
      Quality control filtering threshold
      to exclude from the analysis all
      cells with the percentage of
      transcripts mapped to mitochondrial
      genes exceeding the provided value.
      Default: 5
    "sd:layout":
      advanced: true

  minimum_novelty_score:
    type: string?
    default: "0.8"
    label: "Minimum novelty score per cell"
    doc: |
      Quality control filtering threshold
      to exclude from the analysis all
      cells with the novelty scores
      smaller than the provided value.
      This QC metrics indicates the overall
      transcriptomic dissimilarity of the
      cells and is calculated as the ratio
      of log10(Genes) to log10(RNA UMI).
      If the selected "Cell Ranger ARC
      Sample" includes multiple aggregated
      datasets, each of them can be filtered
      independently by providing comma or
      space-separated list of filtering
      thresholds. The order and number of
      the specified values need to match
      with the datasets order from the
      "aggregation_metadata.csv" output
      generated by "Cell Ranger ARC Sample"
      and accessible on the "Files" tab.
      Default: 0.8
    "sd:layout":
      advanced: true

  minimum_fragments:
    type: string?
    default: "1000"
    label: "Minimum number of fragments in peaks per cell"
    doc: |
      Quality control filtering threshold
      to exclude from the analysis all
      cells with the number of fragments
      in peaks smaller than the provided
      value. If the selected "Cell Ranger
      ARC Sample" includes multiple aggregated
      datasets, each of them can be filtered
      independently by providing comma or
      space-separated list of filtering
      thresholds. The order and number of
      the specified values need to match
      with the datasets order from the
      "aggregation_metadata.csv" output
      generated by "Cell Ranger ARC Sample"
      and accessible on the "Files" tab.
      Default: 1000
    "sd:layout":
      advanced: true

  minimum_tss_enrich:
    type: string?
    default: "2"
    label: "Minimum TSS enrichment score per cell"
    doc: |
      Quality control filtering threshold
      to exclude from the analysis all
      cells with the TSS enrichment score
      smaller than the provided value.
      This QC metrics is calculated based
      on the ratio of ATAC fragments
      centered at the genes TSS to ATAC
      fragments in the TSS-flanking regions.
      If the selected "Cell Ranger ARC
      Sample" includes multiple aggregated
      datasets, each of them can be filtered
      independently by providing comma or
      space-separated list of filtering
      thresholds. The order and number of
      the specified values need to match
      with the datasets order from the
      "aggregation_metadata.csv" output
      generated by "Cell Ranger ARC Sample"
      and accessible on the "Files" tab.
      Default: 2
    "sd:layout":
      advanced: true

  minimum_frip:
    type: string?
    default: "0.15"
    label: "Minimum FRiP per cell"
    doc: |
      Quality control filtering threshold
      to exclude from the analysis all
      cells with the FRiP (Fraction of
      Reads in Peaks) smaller than the
      provided value.
      If the selected "Cell Ranger ARC
      Sample" includes multiple aggregated
      datasets, each of them can be filtered
      independently by providing comma or
      space-separated list of filtering
      thresholds. The order and number of
      the specified values need to match
      with the datasets order from the
      "aggregation_metadata.csv" output
      generated by "Cell Ranger ARC Sample"
      and accessible on the "Files" tab.
      Default: 0.15
    "sd:layout":
      advanced: true

  maximum_nucl_signal:
    type: string?
    default: "4"
    label: "Maximum nucleosome signal per cell"
    doc: |
      Quality control filtering threshold
      to exclude from the analysis all
      cells with the nucleosome signal
      higher than the provided value.
      Nucleosome signal is a measurement
      of nucleosome occupancy. It quantifies
      the approximate ratio of mononucleosomal
      to nucleosome-free ATAC fragments.
      If the selected "Cell Ranger ARC
      Sample" includes multiple aggregated
      datasets, each of them can be filtered
      independently by providing comma or
      space-separated list of filtering
      thresholds. The order and number of
      the specified values need to match
      with the datasets order from the
      "aggregation_metadata.csv" output
      generated by "Cell Ranger ARC Sample"
      and accessible on the "Files" tab.
      Default: 4
    "sd:layout":
      advanced: true

  maximum_blacklist_fraction:
    type: string?
    default: "0.05"
    label: "Maximum blacklist fraction per cell"
    doc: |
      Quality control filtering threshold
      to exclude from the analysis all
      cells with the fraction of ATAC
      fragments in genomic blacklist regions
      bigger than the provided value.
      If the selected "Cell Ranger ARC
      Sample" includes multiple aggregated
      datasets, each of them can be filtered
      independently by providing comma or
      space-separated list of filtering
      thresholds. The order and number of
      the specified values need to match
      with the datasets order from the
      "aggregation_metadata.csv" output
      generated by "Cell Ranger ARC Sample"
      and accessible on the "Files" tab.
      Default: 0.05
    "sd:layout":
      advanced: true

  color_theme:
    type:
    - "null"
    - type: enum
      symbols:
      - "gray"
      - "bw"
      - "linedraw"
      - "light"
      - "dark"
      - "minimal"
      - "classic"
      - "void"
    default: "classic"
    label: "Plots color theme"
    doc: |
      Color theme for all plots saved
      as PNG files.
      Default: classic
    "sd:layout":
      advanced: true

  threads:
    type:
    - "null"
    - type: enum
      symbols:
      - "1"
      - "2"
      - "3"
      - "4"
    default: "1"
    label: "Cores/CPUs"
    doc: |
      Parallelization parameter to define the
      number of cores/CPUs that can be utilized
      simultaneously.
      Default: 1
    "sd:layout":
      advanced: true


outputs:

  raw_1_2_qc_mtrcs_pca_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_1_2_qc_mtrcs_pca_plot_png
    label: "QC metrics PCA (1,2), raw"
    doc: |
      PC1 and PC2 from the QC metrics
      PCA for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "QC metrics PCA (1,2)"

  raw_2_3_qc_mtrcs_pca_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_2_3_qc_mtrcs_pca_plot_png
    label: "QC metrics PCA (2,3), raw"
    doc: |
      PC2 and PC3 from the QC metrics
      PCA for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "QC metrics PCA (2,3)"

  raw_cells_count_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_cells_count_plot_png
    label: "Cells per dataset, raw"
    doc: |
      Number of cells per dataset
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Cells per dataset"

  raw_umi_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_umi_dnst_plot_png
    label: "Transcripts per cell, raw"
    doc: |
      Transcripts per cell density
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Transcripts per cell"

  raw_gene_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_gene_dnst_plot_png
    label: "Genes per cell, raw"
    doc: |
      Genes per cell density
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Genes per cell"

  raw_gene_umi_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_gene_umi_plot_png
    label: "Genes vs transcripts, raw"
    doc: |
      Genes vs transcripts per cell
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Genes vs transcripts"

  raw_mito_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_mito_dnst_plot_png
    label: "Mitochondrial percentage, raw"
    doc: |
      Percentage of transcripts mapped to
      mitochondrial genes per cell density
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Mitochondrial percentage"

  raw_nvlt_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_nvlt_dnst_plot_png
    label: "Novelty score, raw"
    doc: |
      Novelty score per cell density
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Novelty score"

  raw_frgm_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_frgm_dnst_plot_png
    label: "Fragments in peaks per cell, raw"
    doc: |
      Fragments in peaks per cell density
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Fragments in peaks per cell"

  raw_peak_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_peak_dnst_plot_png
    label: "Peaks per cell, raw"
    doc: |
      Peaks per cell density
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Peaks per cell"

  raw_blck_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_blck_dnst_plot_png
    label: "Blacklist regions fraction, raw"
    doc: |
      Fraction of ATAC fragments within
      genomic blacklist regions per cell
      density for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Blacklist regions fraction"

  raw_rna_atac_cnts_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_rna_atac_cnts_plot_png
    label: "Transcripts vs fragments in peaks, raw"
    doc: |
      Transcripts vs fragments in peaks
      per cell for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Transcripts vs fragments in peaks"

  raw_tss_frgm_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_tss_frgm_plot_png
    label: "TSS enrichment score vs fragments in peaks, raw"
    doc: |
      TSS enrichment score vs fragments in peaks
      per cell for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "TSS enrichment score vs fragments in peaks"

  raw_qc_mtrcs_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_qc_mtrcs_dnst_plot_png
    label: "Main QC metrics, raw"
    doc: |
      Main QC metrics per cell densities
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Main QC metrics"

  raw_rnadbl_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_rnadbl_plot_png
    label: "RNA doublets, raw"
    doc: |
      Percentage of RNA doublets per
      dataset for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "RNA doublets"

  raw_atacdbl_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_atacdbl_plot_png
    label: "ATAC doublets, raw"
    doc: |
      Percentage of ATAC doublets per
      dataset for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "ATAC doublets"

  raw_vrlpdbl_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_vrlpdbl_plot_png
    label: "RNA and ATAC doublets overlap, raw"
    doc: |
      RNA and ATAC doublets overlap per
      dataset for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "RNA and ATAC doublets overlap"

  raw_tss_nrch_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_tss_nrch_plot_png
    label: "TSS enrichment, raw"
    doc: |
      TSS enrichment score
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "TSS enrichment"

  raw_frgm_hist_png:
    type: File?
    outputSource: sc_multiome_filter/raw_frgm_hist_png
    label: "Fragments length, raw"
    doc: |
      Fragments length distribution
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw"
        Caption: "Fragments length"

  raw_umi_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_umi_dnst_spl_cnd_plot_png
    label: "Transcripts per cell, raw, split by condition"
    doc: |
      Split by grouping condition transcripts
      per cell density for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw, by condition"
        Caption: "Transcripts per cell"

  raw_gene_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_gene_dnst_spl_cnd_plot_png
    label: "Genes per cell, raw, split by condition"
    doc: |
      Split by grouping condition genes
      per cell for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw, by condition"
        Caption: "Genes per cell"

  raw_mito_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_mito_dnst_spl_cnd_plot_png
    label: "Mitochondrial percentage, raw, split by condition"
    doc: |
      Split by grouping condition the
      percentage of transcripts mapped to
      mitochondrial genes per cell density
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw, by condition"
        Caption: "Mitochondrial percentage"

  raw_nvlt_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_nvlt_dnst_spl_cnd_plot_png
    label: "Novelty score, raw, split by condition"
    doc: |
      Split by grouping condition the
      novelty score per cell density
      for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw, by condition"
        Caption: "Novelty score"

  raw_frgm_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_frgm_dnst_spl_cnd_plot_png
    label: "Fragments in peaks per cell, raw, split by condition"
    doc: |
      Split by grouping condition fragments
      in peaks per cell density for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw, by condition"
        Caption: "Fragments in peaks per cell"

  raw_peak_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_peak_dnst_spl_cnd_plot_png
    label: "Peaks per cell, raw, split by condition"
    doc: |
      Split by grouping condition peaks
      per cell for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw, by condition"
        Caption: "Peaks per cell"

  raw_blck_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/raw_blck_dnst_spl_cnd_plot_png
    label: "Blacklist regions fraction, raw, split by condition"
    doc: |
      Split by grouping condition the
      fraction of ATAC fragments within
      genomic blacklist regions per cell
      density for raw data
    "sd:visualPlugins":
    - image:
        tab: "Raw, by condition"
        Caption: "Blacklist regions fraction"

  fltr_1_2_qc_mtrcs_pca_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_1_2_qc_mtrcs_pca_plot_png
    label: "QC metrics PCA (1,2), filtered"
    doc: |
      PC1 and PC2 from the QC metrics
      PCA for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "QC metrics PCA (1,2)"

  fltr_2_3_qc_mtrcs_pca_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_2_3_qc_mtrcs_pca_plot_png
    label: "QC metrics PCA (2,3), filtered"
    doc: |
      PC2 and PC3 from the QC metrics
      PCA for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "QC metrics PCA (2,3)"

  fltr_cells_count_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_cells_count_plot_png
    label: "Cells per dataset, filtered"
    doc: |
      Number of cells per dataset
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Cells per dataset"

  fltr_umi_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_umi_dnst_plot_png
    label: "Transcripts per cell, filtered"
    doc: |
      Transcripts per cell density
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Transcripts per cell"

  fltr_gene_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_gene_dnst_plot_png
    label: "Genes per cell, filtered"
    doc: |
      Genes per cell density
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Genes per cell"

  fltr_gene_umi_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_gene_umi_plot_png
    label: "Genes vs transcripts, filtered"
    doc: |
      Genes vs transcripts per cell
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Genes vs transcripts"

  fltr_mito_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_mito_dnst_plot_png
    label: "Mitochondrial percentage, filtered"
    doc: |
      Percentage of transcripts mapped to
      mitochondrial genes per cell density
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Mitochondrial percentage"

  fltr_nvlt_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_nvlt_dnst_plot_png
    label: "Novelty score, filtered"
    doc: |
      Novelty score per cell density
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Novelty score"

  fltr_frgm_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_frgm_dnst_plot_png
    label: "Fragments in peaks per cell, filtered"
    doc: |
      Fragments in peaks per cell density
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Fragments in peaks per cell"

  fltr_peak_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_peak_dnst_plot_png
    label: "Peaks per cell, filtered"
    doc: |
      Peaks per cell density
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Peaks per cell"

  fltr_blck_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_blck_dnst_plot_png
    label: "Blacklist regions fraction, filtered"
    doc: |
      Fraction of ATAC fragments within
      genomic blacklist regions per cell
      density for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Blacklist regions fraction"

  fltr_rna_atac_cnts_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_rna_atac_cnts_plot_png
    label: "Transcripts vs fragments in peaks, filtered"
    doc: |
      Transcripts vs fragments in peaks
      per cell for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Transcripts vs fragments in peaks"

  fltr_tss_frgm_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_tss_frgm_plot_png
    label: "TSS enrichment score vs fragments in peaks, filtered"
    doc: |
      TSS enrichment score vs fragments in
      peaks per cell for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "TSS enrichment score vs fragments in peaks"

  fltr_qc_mtrcs_dnst_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_qc_mtrcs_dnst_plot_png
    label: "Main QC metrics, filtered"
    doc: |
      Main QC metrics per cell densities
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Main QC metrics"

  fltr_rnadbl_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_rnadbl_plot_png
    label: "RNA doublets, filtered"
    doc: |
      Percentage of RNA doublets per
      dataset for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "RNA doublets"

  fltr_atacdbl_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_atacdbl_plot_png
    label: "ATAC doublets, filtered"
    doc: |
      Percentage of ATAC doublets per
      dataset for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "ATAC doublets"

  fltr_vrlpdbl_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_vrlpdbl_plot_png
    label: "RNA and ATAC doublets overlap, filtered"
    doc: |
      RNA and ATAC doublets overlap per
      dataset for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "RNA and ATAC doublets overlap"

  fltr_tss_nrch_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_tss_nrch_plot_png
    label: "TSS enrichment, filtered"
    doc: |
      TSS enrichment score
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "TSS enrichment"

  fltr_frgm_hist_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_frgm_hist_png
    label: "Fragments length, filtered"
    doc: |
      Fragments length distribution
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered"
        Caption: "Fragments length"

  fltr_umi_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_umi_dnst_spl_cnd_plot_png
    label: "Transcripts per cell, filtered, split by condition"
    doc: |
      Split by grouping condition transcripts
      per cell density for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered, by condition"
        Caption: "Transcripts per cell"

  fltr_gene_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_gene_dnst_spl_cnd_plot_png
    label: "Genes per cell, filtered, split by condition"
    doc: |
      Split by grouping condition genes
      per cell for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered, by condition"
        Caption: "Genes per cell"

  fltr_mito_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_mito_dnst_spl_cnd_plot_png
    label: "Mitochondrial percentage, filtered, split by condition"
    doc: |
      Split by grouping condition the
      percentage of transcripts mapped to
      mitochondrial genes per cell density
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered, by condition"
        Caption: "Mitochondrial percentage"

  fltr_nvlt_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_nvlt_dnst_spl_cnd_plot_png
    label: "Novelty score, filtered, split by condition"
    doc: |
      Split by grouping condition the
      novelty score per cell density
      for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered, by condition"
        Caption: "Novelty score"

  fltr_frgm_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_frgm_dnst_spl_cnd_plot_png
    label: "Fragments in peaks per cell, filtered, split by condition"
    doc: |
      Split by grouping condition fragments
      in peaks per cell density for filtered
      data
    "sd:visualPlugins":
    - image:
        tab: "Filtered, by condition"
        Caption: "Fragments in peaks per cell"

  fltr_peak_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_peak_dnst_spl_cnd_plot_png
    label: "Peaks per cell, filtered, split by condition"
    doc: |
      Split by grouping condition peaks
      per cell for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered, by condition"
        Caption: "Peaks per cell"

  fltr_blck_dnst_spl_cnd_plot_png:
    type: File?
    outputSource: sc_multiome_filter/fltr_blck_dnst_spl_cnd_plot_png
    label: "Blacklist regions fraction, filtered, split by condition"
    doc: |
      Split by grouping condition the
      fraction of ATAC fragments within
      genomic blacklist regions per cell
      density for filtered data
    "sd:visualPlugins":
    - image:
        tab: "Filtered, by condition"
        Caption: "Blacklist regions fraction"

  ucsc_cb_html_data:
    type: Directory
    outputSource: sc_multiome_filter/ucsc_cb_html_data
    label: "UCSC Cell Browser data"
    doc: |
      Directory with UCSC Cell Browser
      data

  ucsc_cb_html_file:
    type: File
    outputSource: sc_multiome_filter/ucsc_cb_html_file
    label: "UCSC Cell Browser"
    doc: |
      UCSC Cell Browser HTML index file
    "sd:visualPlugins":
    - linkList:
        tab: "Overview"
        target: "_blank"

  seurat_data_rds:
    type: File
    outputSource: sc_multiome_filter/seurat_data_rds
    label: "Processed seurat data in RDS format"
    doc: |
      Processed seurat data in RDS format

  datasets_metadata:
    type: File
    outputSource: sc_multiome_filter/datasets_metadata
    label: "Example of datasets metadata"
    doc: |
      Example of datasets metadata file
      in TSV format

  pdf_plots:
    type: File
    outputSource: compress_pdf_plots/compressed_folder
    label: "Plots in PDF format"
    doc: |
      Compressed folder with plots
      in PDF format

  sc_multiome_filter_stdout_log:
    type: File
    outputSource: sc_multiome_filter/stdout_log
    label: "Output log, filtering step"
    doc: |
      stdout log generated by sc_multiome_filter step

  sc_multiome_filter_stderr_log:
    type: File
    outputSource: sc_multiome_filter/stderr_log
    label: "Error log, filtering step"
    doc: |
      stderr log generated by sc_multiome_filter step


steps:

  uncompress_feature_bc_matrices:
    doc: |
      Extracts the content of TAR file into a folder
    run: ../tools/tar-extract.cwl
    in:
      file_to_extract: filtered_feature_bc_matrix_folder
    out:
    - extracted_folder

  sc_multiome_filter:
    doc: |
      Filters single-cell multiome ATAC and RNA-Seq datasets
      based on the multiple QC metrics
    run: ../tools/sc-multiome-filter.cwl
    in:
      feature_bc_matrices_folder: uncompress_feature_bc_matrices/extracted_folder
      aggregation_metadata: aggregation_metadata
      atac_fragments_file: atac_fragments_file
      annotation_gtf_file: annotation_gtf_file
      chrom_length_file: chrom_length_file
      grouping_data: grouping_data
      blacklist_regions_file: blacklist_regions_file
      barcodes_data: barcodes_data
      rna_minimum_cells:
        default: 1                          # will remove genes that are not expressed in any of the cells
      minimum_genes:
        source: minimum_genes
        valueFrom: $(split_numbers(self))
      maximum_genes:
        source: maximum_genes
        valueFrom: $(split_numbers(self))
      minimum_umis:
        source: minimum_umis
        valueFrom: $(split_numbers(self))
      mito_pattern: mito_pattern
      maximum_mito_perc: maximum_mito_perc
      minimum_novelty_score:
        source: minimum_novelty_score
        valueFrom: $(split_numbers(self))
      atac_minimum_cells:
        default: 1                          # will remove peaks that are not present in any of the cells
      minimum_fragments:
        source: minimum_fragments
        valueFrom: $(split_numbers(self))
      maximum_nucl_signal:
        source: maximum_nucl_signal
        valueFrom: $(split_numbers(self))
      minimum_tss_enrich:
        source: minimum_tss_enrich
        valueFrom: $(split_numbers(self))
      minimum_frip:
        source: minimum_frip
        valueFrom: $(split_numbers(self))
      maximum_blacklist_fraction:
        source: maximum_blacklist_fraction
        valueFrom: $(split_numbers(self))
      remove_doublets:
        source: remove_doublets
        valueFrom: |
          ${
            if (self == "Based on either RNA or ATAC") {
              return "union";
            } else if (self == "Based on RNA") {
              return "onlyrna";
            } else if (self == "Based on ATAC") {
              return "onlyatac";
            } else if (self == "Based on both RNA and ATAC") {
              return "intersect";
            } else {
              return null;
            }
          }
      verbose:
        default: true
      export_ucsc_cb:
        default: true
      export_pdf_plots:
        default: true
      color_theme: color_theme
      parallel_memory_limit:
        default: 32
      vector_memory_limit:
        default: 96
      threads:
        source: threads
        valueFrom: $(parseInt(self))
    out:
    - raw_1_2_qc_mtrcs_pca_plot_png
    - raw_2_3_qc_mtrcs_pca_plot_png
    - raw_cells_count_plot_png
    - raw_umi_dnst_plot_png
    - raw_gene_dnst_plot_png
    - raw_gene_umi_plot_png
    - raw_mito_dnst_plot_png
    - raw_nvlt_dnst_plot_png
    - raw_frgm_dnst_plot_png
    - raw_peak_dnst_plot_png
    - raw_blck_dnst_plot_png
    - raw_rna_atac_cnts_plot_png
    - raw_tss_frgm_plot_png
    - raw_qc_mtrcs_dnst_plot_png
    - raw_rnadbl_plot_png
    - raw_atacdbl_plot_png
    - raw_vrlpdbl_plot_png
    - raw_tss_nrch_plot_png
    - raw_frgm_hist_png
    - raw_umi_dnst_spl_cnd_plot_png
    - raw_gene_dnst_spl_cnd_plot_png
    - raw_mito_dnst_spl_cnd_plot_png
    - raw_nvlt_dnst_spl_cnd_plot_png
    - raw_frgm_dnst_spl_cnd_plot_png
    - raw_peak_dnst_spl_cnd_plot_png
    - raw_blck_dnst_spl_cnd_plot_png
    - fltr_1_2_qc_mtrcs_pca_plot_png
    - fltr_2_3_qc_mtrcs_pca_plot_png
    - fltr_cells_count_plot_png
    - fltr_umi_dnst_plot_png
    - fltr_gene_dnst_plot_png
    - fltr_gene_umi_plot_png
    - fltr_mito_dnst_plot_png
    - fltr_nvlt_dnst_plot_png
    - fltr_frgm_dnst_plot_png
    - fltr_peak_dnst_plot_png
    - fltr_blck_dnst_plot_png
    - fltr_rna_atac_cnts_plot_png
    - fltr_rnadbl_plot_png
    - fltr_atacdbl_plot_png
    - fltr_vrlpdbl_plot_png
    - fltr_tss_frgm_plot_png
    - fltr_qc_mtrcs_dnst_plot_png
    - fltr_tss_nrch_plot_png
    - fltr_frgm_hist_png
    - fltr_umi_dnst_spl_cnd_plot_png
    - fltr_gene_dnst_spl_cnd_plot_png
    - fltr_mito_dnst_spl_cnd_plot_png
    - fltr_nvlt_dnst_spl_cnd_plot_png
    - fltr_frgm_dnst_spl_cnd_plot_png
    - fltr_peak_dnst_spl_cnd_plot_png
    - fltr_blck_dnst_spl_cnd_plot_png
    - raw_1_2_qc_mtrcs_pca_plot_pdf
    - raw_2_3_qc_mtrcs_pca_plot_pdf
    - raw_cells_count_plot_pdf
    - raw_umi_dnst_plot_pdf
    - raw_gene_dnst_plot_pdf
    - raw_gene_umi_plot_pdf
    - raw_mito_dnst_plot_pdf
    - raw_nvlt_dnst_plot_pdf
    - raw_frgm_dnst_plot_pdf
    - raw_peak_dnst_plot_pdf
    - raw_blck_dnst_plot_pdf
    - raw_rna_atac_cnts_plot_pdf
    - raw_tss_frgm_plot_pdf
    - raw_qc_mtrcs_dnst_plot_pdf
    - raw_rnadbl_plot_pdf
    - raw_atacdbl_plot_pdf
    - raw_vrlpdbl_plot_pdf
    - raw_tss_nrch_plot_pdf
    - raw_frgm_hist_pdf
    - raw_umi_dnst_spl_cnd_plot_pdf
    - raw_gene_dnst_spl_cnd_plot_pdf
    - raw_mito_dnst_spl_cnd_plot_pdf
    - raw_nvlt_dnst_spl_cnd_plot_pdf
    - raw_frgm_dnst_spl_cnd_plot_pdf
    - raw_peak_dnst_spl_cnd_plot_pdf
    - raw_blck_dnst_spl_cnd_plot_pdf
    - fltr_1_2_qc_mtrcs_pca_plot_pdf
    - fltr_2_3_qc_mtrcs_pca_plot_pdf
    - fltr_cells_count_plot_pdf
    - fltr_umi_dnst_plot_pdf
    - fltr_gene_dnst_plot_pdf
    - fltr_gene_umi_plot_pdf
    - fltr_mito_dnst_plot_pdf
    - fltr_nvlt_dnst_plot_pdf
    - fltr_frgm_dnst_plot_pdf
    - fltr_peak_dnst_plot_pdf
    - fltr_blck_dnst_plot_pdf
    - fltr_rna_atac_cnts_plot_pdf
    - fltr_rnadbl_plot_pdf
    - fltr_atacdbl_plot_pdf
    - fltr_vrlpdbl_plot_pdf
    - fltr_tss_frgm_plot_pdf
    - fltr_qc_mtrcs_dnst_plot_pdf
    - fltr_tss_nrch_plot_pdf
    - fltr_frgm_hist_pdf
    - fltr_umi_dnst_spl_cnd_plot_pdf
    - fltr_gene_dnst_spl_cnd_plot_pdf
    - fltr_mito_dnst_spl_cnd_plot_pdf
    - fltr_nvlt_dnst_spl_cnd_plot_pdf
    - fltr_frgm_dnst_spl_cnd_plot_pdf
    - fltr_peak_dnst_spl_cnd_plot_pdf
    - fltr_blck_dnst_spl_cnd_plot_pdf
    - ucsc_cb_html_data
    - ucsc_cb_html_file
    - seurat_data_rds
    - datasets_metadata
    - stdout_log
    - stderr_log

  pdf_plots:
    run: ../tools/files-to-folder.cwl
    in:
      input_files:
        source:
        - sc_multiome_filter/raw_1_2_qc_mtrcs_pca_plot_pdf
        - sc_multiome_filter/raw_2_3_qc_mtrcs_pca_plot_pdf
        - sc_multiome_filter/raw_cells_count_plot_pdf
        - sc_multiome_filter/raw_umi_dnst_plot_pdf
        - sc_multiome_filter/raw_gene_dnst_plot_pdf
        - sc_multiome_filter/raw_gene_umi_plot_pdf
        - sc_multiome_filter/raw_mito_dnst_plot_pdf
        - sc_multiome_filter/raw_nvlt_dnst_plot_pdf
        - sc_multiome_filter/raw_frgm_dnst_plot_pdf
        - sc_multiome_filter/raw_peak_dnst_plot_pdf
        - sc_multiome_filter/raw_blck_dnst_plot_pdf
        - sc_multiome_filter/raw_rna_atac_cnts_plot_pdf
        - sc_multiome_filter/raw_tss_frgm_plot_pdf
        - sc_multiome_filter/raw_qc_mtrcs_dnst_plot_pdf
        - sc_multiome_filter/raw_rnadbl_plot_pdf
        - sc_multiome_filter/raw_atacdbl_plot_pdf
        - sc_multiome_filter/raw_vrlpdbl_plot_pdf
        - sc_multiome_filter/raw_tss_nrch_plot_pdf
        - sc_multiome_filter/raw_frgm_hist_pdf
        - sc_multiome_filter/raw_umi_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/raw_gene_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/raw_mito_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/raw_nvlt_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/raw_frgm_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/raw_peak_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/raw_blck_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/fltr_1_2_qc_mtrcs_pca_plot_pdf
        - sc_multiome_filter/fltr_2_3_qc_mtrcs_pca_plot_pdf
        - sc_multiome_filter/fltr_cells_count_plot_pdf
        - sc_multiome_filter/fltr_umi_dnst_plot_pdf
        - sc_multiome_filter/fltr_gene_dnst_plot_pdf
        - sc_multiome_filter/fltr_gene_umi_plot_pdf
        - sc_multiome_filter/fltr_mito_dnst_plot_pdf
        - sc_multiome_filter/fltr_nvlt_dnst_plot_pdf
        - sc_multiome_filter/fltr_frgm_dnst_plot_pdf
        - sc_multiome_filter/fltr_peak_dnst_plot_pdf
        - sc_multiome_filter/fltr_blck_dnst_plot_pdf
        - sc_multiome_filter/fltr_rna_atac_cnts_plot_pdf
        - sc_multiome_filter/fltr_rnadbl_plot_pdf
        - sc_multiome_filter/fltr_atacdbl_plot_pdf
        - sc_multiome_filter/fltr_vrlpdbl_plot_pdf
        - sc_multiome_filter/fltr_tss_frgm_plot_pdf
        - sc_multiome_filter/fltr_qc_mtrcs_dnst_plot_pdf
        - sc_multiome_filter/fltr_tss_nrch_plot_pdf
        - sc_multiome_filter/fltr_frgm_hist_pdf
        - sc_multiome_filter/fltr_umi_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/fltr_gene_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/fltr_mito_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/fltr_nvlt_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/fltr_frgm_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/fltr_peak_dnst_spl_cnd_plot_pdf
        - sc_multiome_filter/fltr_blck_dnst_spl_cnd_plot_pdf
        valueFrom: $(self.flat().filter(n => n))
    out:
    - folder

  compress_pdf_plots:
    run: ../tools/tar-compress.cwl
    in:
      folder_to_compress: pdf_plots/folder
    out:
    - compressed_folder


$namespaces:
  s: http://schema.org/

$schemas:
- https://github.com/schemaorg/schemaorg/raw/main/data/releases/11.01/schemaorg-current-http.rdf

label: "Single-cell Multiome ATAC and RNA-Seq Filtering Analysis"
s:name: "Single-cell Multiome ATAC and RNA-Seq Filtering Analysis"
s:alternateName: "Filters single-cell multiome ATAC and RNA-Seq datasets based on the multiple QC metrics"

s:downloadUrl: https://raw.githubusercontent.com/Barski-lab/workflows-datirium/master/workflows/sc-multiome-filter.cwl
s:codeRepository: https://github.com/Barski-lab/workflows-datirium
s:license: http://www.apache.org/licenses/LICENSE-2.0

s:isPartOf:
  class: s:CreativeWork
  s:name: Common Workflow Language
  s:url: http://commonwl.org/

s:creator:
- class: s:Organization
  s:legalName: "Cincinnati Children's Hospital Medical Center"
  s:location:
  - class: s:PostalAddress
    s:addressCountry: "USA"
    s:addressLocality: "Cincinnati"
    s:addressRegion: "OH"
    s:postalCode: "45229"
    s:streetAddress: "3333 Burnet Ave"
    s:telephone: "+1(513)636-4200"
  s:logo: "https://www.cincinnatichildrens.org/-/media/cincinnati%20childrens/global%20shared/childrens-logo-new.png"
  s:department:
  - class: s:Organization
    s:legalName: "Allergy and Immunology"
    s:department:
    - class: s:Organization
      s:legalName: "Barski Research Lab"
      s:member:
      - class: s:Person
        s:name: Michael Kotliar
        s:email: mailto:misha.kotliar@gmail.com
        s:sameAs:
        - id: http://orcid.org/0000-0002-6486-3898


doc: |
  Single-cell Multiome ATAC and RNA-Seq Filtering Analysis
  
  Filters single-cell multiome ATAC and RNA-Seq datasets
  based on the multiple QC metrics.