# ChIP-Seq basic analysis workflow for single-read data

Reads are aligned to the reference genome with [Bowtie](http://bowtie-bio.sourceforge.net/index.shtml). Results are saved as coordinate sorted [BAM](http://samtools.github.io/hts-specs/SAMv1.pdf) alignment and index BAI files. Optionally, PCR duplicates can be removed. To obtain coverage in [bigWig](https://genome.ucsc.edu/goldenpath/help/bigWig.html) format, average fragment length is calculated by [MACS2](https://github.com/taoliu/MACS), and individual reads are extended to this length in the 3â€™ direction. Areas of enrichment identified by MACS2 are saved in ENCODE [narrow peak](http://genome.ucsc.edu/FAQ/FAQformat.html#format12) or [broad peak](https://genome.ucsc.edu/FAQ/FAQformat.html#format13) formats. Called peaks together with the nearest genes are saved in TSV format. In addition to basic statistics (number of total/mapped/multi-mapped/unmapped/duplicate reads), pipeline generates several quality control measures. Base frequency plots are used to estimate adapter contamination, a frequent occurrence in low-input ChIP-Seq experiments. Expected distinct reads count from [Preseq](http://smithlabresearch.org/software/preseq/) can be used to estimate read redundancy for a given sequencing depth. Average tag density profiles can be used to estimate ChIP enrichment for promoter proximal histone modifications. Use of different parameters for different antibodies (calling broad or narrow peaks) is possible. Additionally, users can elect to use BAM file from another experiment as control for MACS2 peak calling.

## Cite as

*Kartashov AV, Barski A. BioWardrobe: an integrated platform for analysis of epigenomics and transcriptomics data. Genome Biol. 2015;16(1):158. Published 2015 Aug 7. [doi:10.1186/s13059-015-0720-3](https://www.ncbi.nlm.nih.gov/pubmed/26248465)*

## Software versions

- Bowtie 1.2.0
- Samtools 1.4
- Preseq 2.0
- MACS2 2.1.1.20160309
- Bedtools 2.26.0
- UCSC userApps v358

## Inputs

| ID                        | Label                                          | Description                                                                                                                                                      | Required | Default | Upstream analyses               |
| ------------------------- | ---------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------: | ------- | ------------------------------- |
| **fastq\_file**           | FASTQ file                                     | Single-read sequencing data in FASTQ format (fastq, fq, bzip2, gzip, zip)                                                                                        |    +     |         |                                 |
| **indices\_folder**       | Genome indices                                 | Directory with the genome indices generated by Bowtie                                                                                                            |    +     |         | genome\_indices/bowtie\_indices |
| **annotation\_file**      | Genome annotation file                         | Genome annotation file in TSV format                                                                                                                             |    +     |         | genome\_indices/annotation      |
| **genome\_size**          | Effective genome size                          | The length of the mappable genome (hs, mm, ce, dm or number, for example 2.7e9)                                                                                  |    +     |         | genome\_indices/genome\_size    |
| **chrom\_length**         | Chromosome lengths file                        | Chromosome lengths file in TSV format                                                                                                                            |    +     |         | genome\_indices/chrom\_length   |
| **broad\_peak**           | Call broad peaks                               | Make MACS2 call broad peaks by linking nearby highly enriched regions                                                                                            |    +     |         |                                 |
| **control\_file**         | Control ChIP-Seq single-read experiment        | Indexed BAM file from the ChIP-Seq single-read experiment to be used as a control for MACS2 peak calling                                                         |          | Null    | control\_file/bambai\_pair      |
| **exp\_fragment\_size**   | Expected fragment size                         | Expected fragment size for read extenstion towards 3' end if *force\_fragment\_size* was set to True or if calculated by MACS2 fragment size was less that 80 bp |          | 150     |                                 |
| **force\_fragment\_size** | Force peak calling with expected fragment size | Make MACS2 don't build the shifting model and use expected fragment size for read extenstion towards 3' end                                                      |          | False   |                                 |
| **clip\_3p\_end**         | Clip from 3' end                               | Number of base pairs to clip from 3' end                                                                                                                         |          | 0       |                                 |
| **clip\_5p\_end**         | Clip from 5' end                               | Number of base pairs to clip from 5' end                                                                                                                         |          | 0       |                                 |
| **remove\_duplicates**    | Remove PCR duplicates                          | Remove PCR duplicates from sorted BAM file                                                                                                                       |          | False   |                                 |
| **threads**               | Number of threads                              | Number of threads for those steps that support multithreading                                                                                                    |          | 2       |                                 |


## Outputs

| ID                       | Label                              | Description                                                                          | Required | Visualization                                                      |
| ------------------------ | ---------------------------------- | ------------------------------------------------------------------------------------ | :------: | ------------------------------------------------------------------ |
| **fastx\_statistics**    | FASTQ quality statistics           | FASTQ quality statistics in TSV format                                               |    +     | *Base Frequency* and *Quality Control* plots in *QC Plots* tab     |
| **bambai\_pair**         | Aligned reads                      | Coordinate sorted BAM alignment and index BAI files                                  |    +     | *Nucleotide Sequence Alignments* track in *IGV Genome Browser* tab |
| **bigwig**               | Genome coverage                    | Genome coverage in bigWig format                                                     |    +     | *Genome Coverage* track in *IGV Genome Browser* tab                |
| **iaintersect\_result**  | Gene annotated peaks               | MACS2 peak file annotated with nearby genes                                          |    +     | *Peak Coordinates* table in *Peak Calling* tab                     |
| **atdp\_result**         | Average Tag Density Plot           | Average Tag Density Plot file in TSV format                                          |    +     | *Average Tag Density Plot* in *QC Plots* tab                       |
| **macs2\_called\_peaks** | Called peaks                       | Called peaks file with 1-based coordinates in XLS format                             |    +     |                                                                    |
| **macs2\_narrow\_peaks** | Narrow peaks                       | Called peaks file in ENCODE narrow peak format                                       |          | *Narrow peaks* track in *IGV Genome Browser* tab                   |
| **macs2\_broad\_peaks**  | Broad peaks                        | Called peaks file in ENCODE broad peak format                                        |          | *Broad peaks* track in *IGV Genome Browser* tab                    |
| **preseq\_estimates**    | Expected Distinct Reads Count Plot | Expected distinct reads count file from Preseq in TSV format                         |          | *Expected Distinct Reads Count Plot* in *QC Plots* tab             |
| **workflow\_statistics** | Workflow execution statistics      | Overall workflow execution statistics from bowtie\_aligner and samtools\_rmdup steps |    +     | *Overview* tab and experiment's preview                            |
| **bowtie\_log**          | Read alignment log                 | Read alignment log file from Bowtie                                                  |    +     |                                                                    |