cwlVersion: v1.0
class: CommandLineTool


requirements:
- class: InlineJavascriptRequirement
- class: InitialWorkDirRequirement
  listing: |
    ${
      var listing = [
        {
          "entry": inputs.fastq_file_r1,
          "entryname": "sample_S1_L001_R1_001.fastq",
          "writable": true
        },
        {
          "entry": inputs.fastq_file_r2,
          "entryname": "sample_S1_L001_R2_001.fastq",
          "writable": true
        }
      ];
      if (inputs.fastq_file_i1){
        listing.push(
          {
            "entry": inputs.fastq_file_i1,
            "entryname": "sample_S1_L001_I1_001.fastq",
            "writable": true
          }
        );
      };
      return listing;
    }


hints:
- class: DockerRequirement
  dockerPull: cumulusprod/cellranger:8.0.1


inputs:
  
  fastq_file_r1:
    type: File
    doc: |
      FASTQ read 1 file.
      It will be staged into workdir
      as sample_S1_L001_R1_001.fastq.

  fastq_file_r2:
    type: File
    doc: |
      FASTQ read 2 file.
      It will be staged into workdir
      as sample_S1_L001_R2_001.fastq.

  fastq_file_i1:
    type: File?
    doc: |
      FASTQ index file.
      If provided, it will be staged
      into workdir as
      sample_S1_L001_I1_001.fastq.

  indices_folder:
    type: Directory
    inputBinding:
      position: 5
      prefix: "--transcriptome"
    doc: |
      Path of folder containing 10x-compatible
      transcriptome reference. These indices
      should be generated by "cellranger mkref"
      command.

  r1_length:
    type: int?
    inputBinding:
      position: 6
      prefix: "--r1-length"
    doc: |
      Limit the length of the input Read 1 sequence
      of Gene Expression library to the first N bases,
      where N is a user-supplied value. Note that the
      length includes the 10x Barcode and UMI sequences
      so do not set this below 26 for Single Cell 3′ v2
      or Single Cell 5′. This and --r2-length are useful
      options for determining the optimal read length
      for sequencing.

  r2_length:
    type: int?
    inputBinding:
      position: 7
      prefix: "--r2-length"
    doc: |
      Limit the length of the input R2 sequence to the
      first N bases, where N is a user-supplied value.
      Trimming occurs before sequencing metrics are
      computed and therefore, limiting R2 read length
      may affect Q30 scores.

  expect_cells:
    type: int?
    inputBinding:
      position: 8
      prefix: "--expect-cells"
    doc: |
      Expected number of recovered cells.
      Starting in Cell Ranger 7.0, the expected number
      of cells can be either auto-estimated or specified
      with --expect-cells. To replicate an old cellranger
      count analysis, set this parameter to 3,000 cells.

  force_cells:
    type: int?
    inputBinding:
      position: 9
      prefix: "--force-cells"
    doc: |
      Force pipeline to use this number of cells, bypassing
      the cell detection algorithm. Use this if the number
      of cells estimated by Cell Ranger is not consistent
      with the barcode rank plot.

  no_bam:
    type: boolean?
    default: false
    inputBinding:
      prefix: "--create-bam="
      valueFrom: $(self?"false":"true")
      separate: false
      position: 10
    doc: |
      Enable or disable BAM file generation. Setting
      --create-bam=false reduces the total computation
      time and the size of the output directory (BAM
      file not generated). We recommend setting
      --create-bam=true if unsure. See
      https://10xgen.com/create-bam for additional
      guidance [possible values: true, false]

  exclude_introns:
    type: boolean?
    default: false
    inputBinding:
      prefix: "--include-introns="
      valueFrom: $(self?"false":"true")
      separate: false
      position: 11
    doc: |
      Starting from the Cell Ranger v7.0 the intronic reads
      are counted by default for whole transcriptome gene
      expression data, except when --target-panel is used.
      Therefore, here we provide a flag to disable this
      default behavior.

  threads:
    type: int?
    inputBinding:
      position: 12
      prefix: "--localcores"
    doc: |
      Set max cores the pipeline may request at one time.
      Default: all available

  memory_limit:
    type: int?
    inputBinding:
      position: 13
      prefix: "--localmem"
    doc: |
      Set max GB the pipeline may request at one time
      Default: all available

  virt_memory_limit:
    type: int?
    inputBinding:
      position: 14
      prefix: "--localvmem"
    doc: |
      Set max virtual address space in GB for the pipeline
      Default: all available


outputs:

  web_summary_report:
    type: File
    outputBinding:
      glob: "sample/outs/web_summary.html"
    doc: |
      Run summary metrics and charts in HTML format.

  metrics_summary_report:
    type: File
    outputBinding:
      glob: "sample/outs/metrics_summary.csv"
    doc: |
      Run summary metrics in CSV format.

  possorted_genome_bam_bai:
    type: File?
    outputBinding:
      glob: "sample/outs/possorted_genome_bam.bam"
    secondaryFiles:
    - .bai
    doc: |
      Indexed reads aligned to the genome
      and transcriptome annotated with
      barcode information.
  
  filtered_feature_bc_matrix_folder:
    type: Directory
    outputBinding:
      glob: "sample/outs/filtered_feature_bc_matrix"
    doc: |
      Folder with filtered feature-barcode matrices
      containing only cellular barcodes in MEX format.

  filtered_feature_bc_matrix_h5:
    type: File
    outputBinding:
      glob: "sample/outs/filtered_feature_bc_matrix.h5"
    doc: |
      Filtered feature-barcode matrices containing
      only cellular barcodes in HDF5 format.
  
  raw_feature_bc_matrices_folder:
    type: Directory
    outputBinding:
      glob: "sample/outs/raw_feature_bc_matrix"
    doc: |
      Folder with unfiltered feature-barcode matrices
      containing all barcodes in MEX format.

  raw_feature_bc_matrices_h5:
    type: File
    outputBinding:
      glob: "sample/outs/raw_feature_bc_matrix.h5"
    doc: |
      Unfiltered feature-barcode matrices containing
      all barcodes in HDF5 format.

  secondary_analysis_report_folder:
    type: Directory
    outputBinding:
      glob: "sample/outs/analysis"
    doc: |
      Folder with secondary analysis results
      including dimensionality reduction, cell
      clustering, and differential expression.

  molecule_info_h5:
    type: File
    outputBinding:
      glob: "sample/outs/molecule_info.h5"
    doc: |
      Molecule-level information used by cellranger
      aggr to aggregate samples into larger datasets.

  loupe_browser_track:
    type: File
    outputBinding:
      glob: "sample/outs/cloupe.cloupe"
    doc: |
      Loupe Browser visualization and
      analysis file.

  stdout_log:
    type: stdout

  stderr_log:
    type: stderr


baseCommand: ["cellranger", "count", "--disable-ui", "--fastqs", ".", "--sample", "sample", "--id", "sample"]


stdout: cellranger_count_stdout.log
stderr: cellranger_count_stderr.log


$namespaces:
  s: http://schema.org/

$schemas:
- https://github.com/schemaorg/schemaorg/raw/main/data/releases/11.01/schemaorg-current-http.rdf

s:name: "Cell Ranger Count (RNA)"
label: "Cell Ranger Count (RNA)"
s:alternateName: "Quantifies single-cell gene expression of the sequencing data from a single 10x Genomics library"

s:downloadUrl: https://raw.githubusercontent.com/Barski-lab/workflows/master/tools/cellranger-count.cwl
s:codeRepository: https://github.com/Barski-lab/workflows
s:license: http://www.apache.org/licenses/LICENSE-2.0

s:isPartOf:
  class: s:CreativeWork
  s:name: Common Workflow Language
  s:url: http://commonwl.org/

s:creator:
- class: s:Organization
  s:legalName: "Cincinnati Children's Hospital Medical Center"
  s:location:
  - class: s:PostalAddress
    s:addressCountry: "USA"
    s:addressLocality: "Cincinnati"
    s:addressRegion: "OH"
    s:postalCode: "45229"
    s:streetAddress: "3333 Burnet Ave"
    s:telephone: "+1(513)636-4200"
  s:logo: "https://www.cincinnatichildrens.org/-/media/cincinnati%20childrens/global%20shared/childrens-logo-new.png"
  s:department:
  - class: s:Organization
    s:legalName: "Allergy and Immunology"
    s:department:
    - class: s:Organization
      s:legalName: "Barski Research Lab"
      s:member:
      - class: s:Person
        s:name: Michael Kotliar
        s:email: mailto:misha.kotliar@gmail.com
        s:sameAs:
        - id: http://orcid.org/0000-0002-6486-3898


doc: |
  Cell Ranger Count (RNA)

  Quantifies single-cell gene expression of the sequencing
  data from a single 10x Genomics library.

  Starting from the Cell Ranger v7.0 the intronic reads are
  counted by default for whole transcriptome gene expression
  data. For more details see
  https://support.10xgenomics.com/docs/intron-mode-rec

  Input parameters for Feature Barcode, Targeted Gene Expression
  analyses are not implemented, therefore the correspondent outputs
  are also excluded.

  Parameters set by default:
  --disable-ui - no need in any UI when running in Docker container
  --id         - can be hardcoded as we rename input files anyway
  --fastqs     - points to the current directory, because input
                 FASTQ files are staged there
  --sample     - hardcoded to sample as we stage input fastq files
                 with the hardcoded names

  Not implemented parameters:
  --description                   - not needed for now
  --project                       - no needed to select input files by folder
  --lanes                         - not needed for now
  --libraries                     - needed only for Gene expression + Feature Barcode analysis
  --feature-ref                   - needed only for Feature Barcode analysis
  --target-panel                  - needed only for Targeted Gene Expression analysis
  --nosecondary                   - no reason to disable it
  --chemistry                     - cell ranger will autodetect the library by default
  --no-libraries                  - used only in Feature Barcode analysis
  --check-library-compatibility   - no reason to disable it
  --min-crispr-umi                - needed only for Protospacer calling (for pooled CRISPR screens)
  --no-target-umi-filter          - needed only for Targeted Gene Expression analysis
  --dry                           - not applicable to our use case
  --jobmode                       - we use default local mode
  --mempercore                    - not used for local mode
  --maxjobs                       - not used for local mode
  --jobinterval                   - not used for local mode
  --overrides                     - not needed for now
  --uiport                        - we disabled UI
  --noexit                        - we disabled UI
  --nopreflight                   - no reason to skip preflight checks
  --output-dir                    - will be saved to the sample/outs by default


s:about: |
  Count gene expression and/or feature barcode reads from a single sample and GEM well

  Usage: cellranger count [OPTIONS] --id <ID> --transcriptome <PATH> --create-bam <true|false>

  Options:
        --id <ID>                                   A unique run id and output folder name [a-zA-Z0-9_-]+
        --description <TEXT>                        Sample description to embed in output files [default: ]
        --transcriptome <PATH>                      Path of folder containing 10x-compatible transcriptome reference
        --fastqs <PATH>                             Path to input FASTQ data
        --project <TEXT>                            Name of the project folder within a mkfastq or bcl2fastq-generated folder from which to pick FASTQs
        --sample <PREFIX>                           Prefix of the filenames of FASTQs to select
        --lanes <NUMS>                              Only use FASTQs from selected lanes
        --libraries <CSV>                           CSV file declaring input library data sources
        --feature-ref <CSV>                         Feature reference CSV file, declaring Feature Barcode constructs and associated barcodes
        --expect-cells <NUM>                        Expected number of recovered cells, used as input to cell calling algorithm
        --force-cells <NUM>                         Force pipeline to use this number of cells, bypassing cell calling algorithm. [MINIMUM: 10]
        --create-bam <true|false>                   Enable or disable BAM file generation. Setting --create-bam=false reduces the total computation time and the size of the
                                                    output directory (BAM file not generated). We recommend setting --create-bam=true if unsure. See
                                                    https://10xgen.com/create-bam for additional guidance [possible values: true, false]
        --nosecondary                               Disable secondary analysis, e.g. clustering. Optional
        --r1-length <NUM>                           Hard trim the input Read 1 to this length before analysis
        --r2-length <NUM>                           Hard trim the input Read 2 to this length before analysis
        --include-introns <true|false>              Include intronic reads in count [default: true] [possible values: true, false]
        --chemistry <CHEM>                          Assay configuration. NOTE: by default the assay configuration is detected automatically, which is the recommended mode.
                                                    You usually will not need to specify a chemistry. Options are: 'auto' for autodetection, 'threeprime' for Single Cell
                                                    3', 'fiveprime' for  Single Cell 5', 'SC3Pv1' or 'SC3Pv2' or 'SC3Pv3' or 'SC3Pv4' for Single Cell 3' v1/v2/v3/v4,
                                                    'SC3Pv3LT' for Single Cell 3' v3 LT, 'SC3Pv3HT' for Single Cell 3' v3 HT, 'SC5P-PE' or 'SC5P-PE-v3' or 'SC5P-R2' or
                                                    'SC5P-R2-v3', for Single Cell 5', paired-end/R2-only, 'SC-FB' for Single Cell Antibody-only 3' v2 or 5'. To analyze the
                                                    GEX portion of multiome data, chemistry must be set to 'ARC-v1' [default: auto]
        --no-libraries                              Proceed with processing using a --feature-ref but no Feature Barcode libraries specified with the 'libraries' flag
        --check-library-compatibility <true|false>  Whether to check for barcode compatibility between libraries. [default: true] [possible values: true, false]
        --min-crispr-umi <NUM>                      Minimum CRISPR UMI threshold [default: 3]
        --dry                                       Do not execute the pipeline. Generate a pipeline invocation (.mro) file and stop
        --jobmode <MODE>                            Job manager to use. Valid options: local (default), sge, lsf, slurm or path to a .template file. Search for help on
                                                    "Cluster Mode" at support.10xgenomics.com for more details on configuring the pipeline to use a compute cluster
        --localcores <NUM>                          Set max cores the pipeline may request at one time. Only applies to local jobs
        --localmem <NUM>                            Set max GB the pipeline may request at one time. Only applies to local jobs
        --localvmem <NUM>                           Set max virtual address space in GB for the pipeline. Only applies to local jobs
        --mempercore <NUM>                          Reserve enough threads for each job to ensure enough memory will be available, assuming each core on your cluster has at
                                                    least this much memory available. Only applies to cluster jobmodes
        --maxjobs <NUM>                             Set max jobs submitted to cluster at one time. Only applies to cluster jobmodes
        --jobinterval <NUM>                         Set delay between submitting jobs to cluster, in ms. Only applies to cluster jobmodes
        --overrides <PATH>                          The path to a JSON file that specifies stage-level overrides for cores and memory. Finer-grained than --localcores,
                                                    --mempercore and --localmem. Consult https://support.10xgenomics.com/ for an example override file
        --output-dir <PATH>                         Output the results to this directory
        --uiport <PORT>                             Serve web UI at http://localhost:PORT
        --disable-ui                                Do not serve the web UI
        --noexit                                    Keep web UI running after pipestance completes or fails
        --nopreflight                               Skip preflight checks
    -h, --help                                      Print help