cwlVersion: v1.0
class: CommandLineTool


requirements:
- class: InlineJavascriptRequirement
- class: InitialWorkDirRequirement
  listing: |
    ${
      var listing = [
        {
          "entry": inputs.fastq_file_r1,
          "entryname": "sample_S1_L001_R1_001.fastq",
          "writable": true
        },
        {
          "entry": inputs.fastq_file_r2,
          "entryname": "sample_S1_L001_R2_001.fastq",
          "writable": true
        },
        {
          "entry": inputs.fastq_file_r3,
          "entryname": "sample_S1_L001_R3_001.fastq",
          "writable": true
        }
      ];
      if (inputs.fastq_file_i1){
        listing.push(
          {
            "entry": inputs.fastq_file_i1,
            "entryname": "sample_S1_L001_I1_001.fastq",
            "writable": true
          }
        );
      };
      return listing;
    }


hints:
- class: DockerRequirement
  dockerPull: cumulusprod/cellranger-atac:2.1.0


inputs:
  
  fastq_file_r1:
    type: File
    doc: |
      FASTQ read 1 file.
      It will be staged into workdir
      as sample_S1_L001_R1_001.fastq.

  fastq_file_r2:
    type: File
    doc: |
      FASTQ read 2 file.
      It will be staged into workdir
      as sample_S1_L001_R2_001.fastq.

  fastq_file_r3:
    type: File
    doc: |
      FASTQ read 3 file.
      It will be staged into workdir
      as sample_S1_L001_R3_001.fastq.

  fastq_file_i1:
    type: File?
    doc: |
      FASTQ index file.
      It will be staged into workdir
      as sample_S1_L001_I1_001.fastq.

  indices_folder:
    type: Directory
    inputBinding:
      position: 10
      prefix: "--reference"
    doc: |
      Path to folder containing a Cell
      Ranger ATAC or Cell Ranger ARC
      reference. Should be generated by
      "cellranger-atac mkref" or
      "cellranger-arc mkref" commands.

  force_cells:
    type: int?
    inputBinding:
      position: 11
      prefix: "--force-cells"
    doc: |
      Define the top N barcodes with the
      most fragments overlapping peaks as
      cells. N must be a positive integer
      <= 20,000. Please consult the
      documentation before using this option.

  chemistry:
    type: string?
    inputBinding:
      position: 12
      prefix: "--chemistry"
    doc: |
      Assay configuration. NOTE: by default
      the assay configuration is detected
      automatically. Use "ARC-v1" to indicate
      that it is a library from the multiome
      assay.

  threads:
    type: int?
    inputBinding:
      position: 13
      prefix: "--localcores"
    doc: |
      Set max cores the pipeline
      may request at one time.
      Default: all available

  memory_limit:
    type: int?
    inputBinding:
      position: 14
      prefix: "--localmem"
    doc: |
      Set max GB the pipeline
      may request at one time.
      Default: all available

  virt_memory_limit:
    type: int?
    inputBinding:
      position: 15
      prefix: "--localvmem"
    doc: |
      Set max virtual address space
      in GB for the pipeline.
      Default: all available


outputs:

  web_summary_report:
    type: File
    outputBinding:
      glob: "sample/outs/web_summary.html"
    doc: |
      Run summary metrics and charts
      in HTML format.

  metrics_summary_report_json:
    type: File
    outputBinding:
      glob: "sample/outs/summary.json"
    doc: |
      Run summary metrics
      in JSON format.

  metrics_summary_report_csv:
    type: File
    outputBinding:
      glob: "sample/outs/summary.csv"
    doc: |
      Run summary metrics
      in CSV format.

  barcode_metrics_report:
    type: File
    outputBinding:
      glob: "sample/outs/singlecell.csv"
    doc: |
      Per-barcode fragment counts &
      metrics in CSV format.

  possorted_genome_bam_bai:
    type: File?
    outputBinding:
      glob: "sample/outs/possorted_bam.bam"
    secondaryFiles:
    - .bai
    doc: |
      Indexed position-sorted reads
      aligned to the genome annotated
      with barcode information in
      BAM format.

  fragments_file:
    type: File
    outputBinding:
      glob: "sample/outs/fragments.tsv.gz"
    secondaryFiles:
    - .tbi
    doc: |
      Count and barcode information for
      every ATAC fragment observed
      in the experiment in TSV format.

  peaks_bed_file:
    type: File
    outputBinding:
      glob: "sample/outs/peaks.bed"
    doc: |
      Locations of open-chromatin regions
      identified in the experiment (these
      regions are referred to as "peaks").

  peak_annotation_file:
    type: File
    outputBinding:
      glob: "sample/outs/peak_annotation.tsv"
    doc: |
      Annotations of peaks based on
      genomic proximity alone.

  cut_sites_bigwig_file:
    type: File
    outputBinding:
      glob: "sample/outs/cut_sites.bigwig"
    doc: |
      Smoothed transposition site track
      in bigWig format.

  peak_motif_mapping_bed:
    type: File?
    outputBinding:
      glob: "sample/outs/peak_motif_mapping.bed"
    doc: |
      File with peak-motif associations
      in BED format.

  filtered_feature_bc_matrix_folder:
    type: Directory
    outputBinding:
      glob: "sample/outs/filtered_peak_bc_matrix"
    doc: |
      Folder with filtered peak-barcode
      matrices containing only cellular
      barcodes in MEX format.

  filtered_feature_bc_matrix_h5:
    type: File
    outputBinding:
      glob: "sample/outs/filtered_peak_bc_matrix.h5"
    doc: |
      Filtered peak-barcode matrices
      containing only cellular barcodes
      in HDF5 format.

  filtered_tf_bc_matrix_folder:
    type: Directory?
    outputBinding:
      glob: "sample/outs/filtered_tf_bc_matrix"
    doc: |
      Folder with filtered tf-barcode
      matrices containing only cellular
      barcodes in MEX format.

  filtered_tf_bc_matrix_h5:
    type: File?
    outputBinding:
      glob: "sample/outs/filtered_tf_bc_matrix.h5"
    doc: |
      Filtered tf-barcode matrices containing
      only cellular barcodes in HDF5 format.

  raw_feature_bc_matrices_folder:
    type: Directory
    outputBinding:
      glob: "sample/outs/raw_peak_bc_matrix"
    doc: |
      Folder with unfiltered peak-barcode
      matrices containing all barcodes
      in MEX format.

  raw_feature_bc_matrices_h5:
    type: File
    outputBinding:
      glob: "sample/outs/raw_peak_bc_matrix.h5"
    doc: |
      Unfiltered peak-barcode matrices
      containing all barcodes in HDF5 format.

  secondary_analysis_report_folder:
    type: Directory
    outputBinding:
      glob: "sample/outs/analysis"
    doc: |
      Folder with secondary
      analysis results.

  loupe_browser_track:
    type: File
    outputBinding:
      glob: "sample/outs/cloupe.cloupe"
    doc: |
      Loupe Browser visualization
      and analysis file.

  stdout_log:
    type: stdout

  stderr_log:
    type: stderr


baseCommand: ["cellranger-atac", "count", "--disable-ui", "--fastqs", ".", "--sample", "sample", "--id", "sample"]


stdout: cellranger_atac_count_stdout.log
stderr: cellranger_atac_count_stderr.log


$namespaces:
  s: http://schema.org/

$schemas:
- https://github.com/schemaorg/schemaorg/raw/main/data/releases/11.01/schemaorg-current-http.rdf

label: "Cell Ranger Count (ATAC)"
s:name: "Cell Ranger Count (ATAC)"
s:alternateName: "Cell Ranger Count (ATAC)"

s:downloadUrl: https://raw.githubusercontent.com/Barski-lab/workflows/master/tools/cellranger-atac-count.cwl
s:codeRepository: https://github.com/Barski-lab/workflows
s:license: http://www.apache.org/licenses/LICENSE-2.0

s:isPartOf:
  class: s:CreativeWork
  s:name: Common Workflow Language
  s:url: http://commonwl.org/

s:creator:
- class: s:Organization
  s:legalName: "Cincinnati Children's Hospital Medical Center"
  s:location:
  - class: s:PostalAddress
    s:addressCountry: "USA"
    s:addressLocality: "Cincinnati"
    s:addressRegion: "OH"
    s:postalCode: "45229"
    s:streetAddress: "3333 Burnet Ave"
    s:telephone: "+1(513)636-4200"
  s:logo: "https://www.cincinnatichildrens.org/-/media/cincinnati%20childrens/global%20shared/childrens-logo-new.png"
  s:department:
  - class: s:Organization
    s:legalName: "Allergy and Immunology"
    s:department:
    - class: s:Organization
      s:legalName: "Barski Research Lab"
      s:member:
      - class: s:Person
        s:name: Michael Kotliar
        s:email: mailto:misha.kotliar@gmail.com
        s:sameAs:
        - id: http://orcid.org/0000-0002-6486-3898


doc: |
  Cell Ranger Count (ATAC)

  Quantifies single-cell chromatin accessibility of the
  sequencing data from a single 10x Genomics library.

  Parameters set by default:
  --disable-ui - no need in any UI when running in Docker container
  --id         - can be hardcoded as we rename input files anyway
  --fastqs     - points to the current directory, because input
                 FASTQ files are staged there
  --sample     - hardcoded to sample as we stage input fastq files
                 with the hardcoded names

  Not implemented parameters:
  --description                   - not needed for now
  --project                       - no needed to select input files by folder
  --lanes                         - not needed for now
  --peaks                         - not needed for now
  --dim-reduce                    - not needed for now
  --subsample-rate                - not needed for now
  --dry                           - not applicable to our use case
  --jobmode                       - we use default local mode
  --mempercore                    - not used for local mode
  --maxjobs                       - not used for local mode
  --jobinterval                   - not used for local mode
  --overrides                     - not needed for now
  --uiport                        - we disabled UI
  --noexit                        - we disabled UI
  --nopreflight                   - no reason to skip preflight checks


s:about: |
  USAGE:
      cellranger-atac count [OPTIONS] --id <ID> --reference <PATH> --fastqs <PATH>

  OPTIONS:
          --id <ID>                      A unique run id and output folder name [a-zA-Z0-9_-]+ of maximum length 64 characters
          --description <TEXT>           Sample description to embed in output files [default: ]
          --reference <PATH>             Path to folder containing a Cell Ranger ATAC or Cell Ranger ARC reference
          --fastqs <PATH>                Path to input FASTQ data
          --project <TEXT>               Name of the project folder within a mkfastq or bcl2fastq-generated folder to pick FASTQs from
          --sample <PREFIX>              Prefix of the filenames of FASTQs to select
          --lanes <NUMS>                 Only use FASTQs from selected lanes
          --force-cells <NUM>            Define the top N barcodes with the most fragments overlapping peaks as cells. N must be a positive integer <= 20,000. Please consult the
                                        documentation before using this option
          --peaks <BED>                  Override peak caller: specify peaks to use in downstream analyses from supplied 3-column BED file. The supplied peaks file must be sorted by
                                        position and not contain overlapping peaks; comment lines beginning with `#` are allowed
          --dim-reduce <STR>             Dimensionality reduction mode for clustering [default: lsa] [possible values: lsa, pca, plsa]
          --subsample-rate <FRACTION>    Downsample to preserve this fraction of reads
          --dry                          Do not execute the pipeline. Generate a pipeline invocation (.mro) file and stop
          --jobmode <MODE>               Job manager to use. Valid options: local (default), sge, lsf, slurm or path to a .template file. Search for help on "Cluster Mode" at
                                        support.10xgenomics.com for more details on configuring the pipeline to use a compute cluster [default: local]
          --localcores <NUM>             Set max cores the pipeline may request at one time. Only applies to local jobs
          --localmem <NUM>               Set max GB the pipeline may request at one time. Only applies to local jobs
          --localvmem <NUM>              Set max virtual address space in GB for the pipeline. Only applies to local jobs
          --mempercore <NUM>             Reserve enough threads for each job to ensure enough memory will be available, assuming each core on your cluster has at least this much
                                        memory available. Only applies to cluster jobmodes
          --maxjobs <NUM>                Set max jobs submitted to cluster at one time. Only applies to cluster jobmodes
          --jobinterval <NUM>            Set delay between submitting jobs to cluster, in ms. Only applies to cluster jobmodes
          --overrides <PATH>             The path to a JSON file that specifies stage-level overrides for cores and memory. Finer-grained than --localcores, --mempercore and
                                        --localmem. Consult https://support.10xgenomics.com/ for an example override file
          --uiport <PORT>                Serve web UI at http://localhost:PORT
          --disable-ui                   Do not serve the web UI
          --noexit                       Keep web UI running after pipestance completes or fails
          --nopreflight                  Skip preflight checks
          --help                         Print help information